File: Top-Interview-Questions/Strings/344-reverse-string.py
class Solution:
    # Sol 1. Reverse built-in func
    def reverseString(self, s: List[str]) -> None:
        """
        Do not return anything, modify s in-place instead.
        """
        s.reverse()
    # Time Complexity: O(n), where n is the length of the list
    # Space Complexity: O(1)
    # Runtime Run Code: 28 ms
    # Submit Runtime: 360 ms, 7.17% faster
    # Submit Memory: 18.3 MB, 5.81% less

    # Sol 2. Swapping using two pointers
    def reverseString(self, s: List[str]) -> None:
        """
        Do not return anything, modify s in-place instead.
        """
        # Edge Case
        if len(s)<=1:
            return
        
        left_pointer = 0
        right_pointer = len(s) - 1
        while left_pointer < right_pointer:
            s[left_pointer], s[right_pointer ] = s[right_pointer], s[left_pointer] # Swapping
            left_pointer += 1
            right_pointer -= 1
    # Time Complexity: O(n)
    # Space Complexity: O(1)
    # Runtime Run Code: 40 ms
    # Submit Runtime: 372 ms, 6.34% faster
    # Submit Memory: 18.3 mb, 5.81% less

    # Sol 3. Even faster
    def reverseString(self, s: List[str]) -> None:
        """
        Do not return anything, modify s in-place instead.
        """
        # Both working
        # n = len(s)
        # for i in range(int(n/2)):
        #     tmp = s[i]
        #     s[i] = s[n-1-i]
        #     s[n-1-i] = tmp
        s[::] = s[::-1]

File: Top-Interview-Questions/Strings-Easy-Collection/344-Reverse_String.cpp
// Time Complexity: O(n/2) => O(n)
// Space Complexity: O(1) in-place by swapping
// Runtime: 52 ms, faster than 79.86% 
// Memory Usage: 23.5 MB, less than 6.10%
class Solution {
public:
    void reverseString(vector<char>& s) {
        
        if(s.size() == 0 || s.size() == 1)
            return;
        
        for(int i = 0; i < s.size()/2; i++){
            swap(s[i],s[s.size()-i-1]);
        }
    }
};